### ISSUE TITLE
Spanner.Data 2-Beta-5 - C# - Nullable Timestamp is dropping results without an error from a select query results.

### ISSUE THREAD

**Issue by alyaros** [googleapis/google-cloud-dotnet#2779]
Running the following scalar query on the web console or C# driver 2-beta-5 is resulting 476577 results.
SELECT count(*) FROM Players;

Running the following query with C# Spanner.Data 2-beta-5 is resulting 452593 results.
SELECT p.PlayerId, p.FirstRealPurchaseTime FROM Players As p;

When I removed the p.FirstRealPurchaseTime which is a Nullable Timestamp and ran:
SELECT p.PlayerId FROM Players As p;
I got the correct number of records which is 476577.

Meaning that when a Nullable Timestamp field (FirstRealPurchaseTime) is presented some records are not returning and no error get thrown in the driver side.

Example Code:

```csharp
var players = new List<string>();
using (var connection = 
new SpannerConnection("Data source=projects/XX/instances/YY/databases/ZZ"))            
{
   connection.Open();
   var command = connection.
                   CreateSelectCommand("SELECT p.PlayerId, p.FirstRealPurchaseTime FROM Players AS p");  
   var result = command.ExecuteReaderAsync().Result;                 
   if (result.HasRows)
   {
                     while (result.Read()) {
                        players.Add(result[0].ToString());  
                       }
    }
    connection.Close();
}
Console.WriteLine("Query Returned " + players.Count); //452593
Console.WriteLine("Query Retunred On Google Web Console  476577 Results");
```

if you remove "p.FirstRealPurchaseTime" from the select query all rows will return as expected.

**Comment by jskeet** [comment]
Thanks for the report. I'll investigate as soon as I'm back from holidays, which will be January 2nd.

In the meantime, could I just check:

- Is this consistent, as in you get the same results every time?
- Is the difference between the two (i.e. 452593 - 476577 = 23984) the same as the number of records where FirstRealPurchaseTime is null?

**Comment by alyaros** [comment]
Consistent.
The 20k diff is not null or not null, it's a mix of both.

**Comment by jskeet** [comment]
Just to check, you mentioned the version number in the title - I wouldn't expect this to differ based on whether you're using 2.0.0-beta05 or an earlier version. Could you let me know if you *are* seeing any difference between versions?

I'm going to try to reproduce this today, but it's possible that I'll need more information about the data. I'll update the bug when I've seen or not seen it.

**Comment by alyaros** [comment]
Hey,
In previous versions we haven't encountered this issue.(2.0.0 Beta 03) and
it doesn't reproducible there.
Thanks

**Comment by jskeet** [comment]
Okay, that's pretty surprising, but thanks for the info. I'm populating a table now (with the same schema as #2777 so that I can hopefully use one table for both issues).

If you're able to test the same query with 2.0.0-beta03 again (in case it's service or data changes that happened at roughly the same time as your update to 2.0.0-beta05) that would be great, but if you don't have a test program handy that would easily demonstrate it, don't worry.

**Comment by jskeet** [comment]
I haven't been able to reproduce this. Complete sample code is included below - you just need to provide your data source and a table name as command line arguments. (I'm assuming that default credentials are okay.) I've built this against 2.0.0-beta05.

You can easily run the test against existing data by commenting out the calls to `CreateTable` and `PopulateTable`.

Could you try this code against your Spanner instance and report the results? It's going to be very hard to work out what's going on if we can't reproduce it :(

```csharp
using Google.Cloud.Spanner.Data;
using System;

namespace Issue2779
{
    class Program
    {
        private const int RowCount = 500_000;
        private const int BatchSize = 2000;
        private const int BatchCount = RowCount / BatchSize;
        private const int NullTimestampPercentage = 10;
        private const int BatchReportingFrequency = 10;

        static void Main(string[] args)
        {
            if (args.Length != 2)
            {
                Console.WriteLine("Please specify 2 command line arguments:");
                Console.WriteLine("- Data source (projects/x/instances/y/databases/z)");
                Console.WriteLine("- Table name");
                return;
            }

            string connectionString = $"Data Source={args[0]}";
            string tableName = args[1];

            CreateTable(connectionString, tableName);
            PopulateTable(connectionString, tableName);
            CountQueryResults(connectionString, $"SELECT p.PlayerId FROM {tableName} AS p");
            CountQueryResults(connectionString, $"SELECT p.PlayerId, p.FirstRealPurchaseTime FROM {tableName} AS p");
        }

        static void CreateTable(string connectionString, string tableName)
        {
            Log("Creating table");
            using (var connection = new SpannerConnection(connectionString))
            {
                string sql = $@"CREATE TABLE {tableName} (
                                PlayerId STRING(MAX) NOT NULL,
                                FirstRealPurchaseTime TIMESTAMP,
                                IsRandom BOOL,
                                RandomNumber INT64
                                ) PRIMARY KEY (PlayerId)";

                using (var command = connection.CreateDdlCommand(sql))
                {
                    command.ExecuteNonQuery();
                }
            }
        }

        static void PopulateTable(string connectionString, string tableName)
        {
            Log("Populating table");
            DateTime timestamp = new DateTime(2000, 1, 1, 0, 0, 0, DateTimeKind.Utc);

            Random random = new Random();

            using (var connection = new SpannerConnection(connectionString))
            {
                var cmd = connection.CreateInsertCommand(tableName);
                var idParameter = cmd.Parameters.Add("PlayerId", SpannerDbType.String);
                var timestampParameter = cmd.Parameters.Add("FirstRealPurchaseTime", SpannerDbType.Timestamp);
                var isRandomParameter = cmd.Parameters.Add("IsRandom", SpannerDbType.Bool);
                var randomNumberParameter = cmd.Parameters.Add("RandomNumber", SpannerDbType.Int64);
                for (int i = 0; i < BatchCount; i++)
                {
                    if (i % BatchReportingFrequency == 0)
                    {
                        Log($"Populating batch {i}");
                    }

                    // Note: normally this code would be inside a retry loop. I've omitted that for simplicity -
                    // I just rerun the whole test if it fails.
                    using (var transaction = connection.BeginTransaction())
                    {
                        cmd.Transaction = transaction;
                        for (int j = 0; j < BatchSize; j++)
                        {
                            idParameter.Value = Guid.NewGuid().ToString();
                            timestampParameter.Value = random.Next(100) < NullTimestampPercentage ? null : (object) timestamp;
                            isRandomParameter.Value = random.Next(2) == 1;
                            randomNumberParameter.Value = random.Next();
                            cmd.ExecuteNonQuery();
                            timestamp = timestamp.AddMinutes(1);
                        }
                        transaction.Commit();
                    }
                }
            }
            Log($"Completed populating table");
        }

        static void CountQueryResults(string connectionString, string sql)
        {
            Log($"Counting results for {sql}");
            using (var connection = new SpannerConnection(connectionString))
            {
                var command = connection.CreateSelectCommand(sql);
                using (var result = command.ExecuteReader())
                {
                    int count = 0;
                    while (result.Read())
                    {
                        count++;
                    }
                    Log($"Total rows: {count}");
                }
            }
        }

        static void Log(string message) =>
            Console.WriteLine(FormattableString.Invariant($"{DateTime.UtcNow:HH:mm:ss'Z'} {message}"));
    }
}
```

**Comment by alyaros** [comment]
Hey,
I've tried to reproduce it over my data table.
Everything runs on the main thread (Single Thread)
Correct Count is = 476577.

Program:
 CountQueryResults(connectionString, $"SELECT p.PlayerId FROM {tableName} AS p");
            CountQueryResults(connectionString, $"SELECT p.PlayerId, p.FirstRealPurchaseTime FROM {tableName} AS p");
            CountQueryResults(connectionString, $"SELECT p.PlayerId, p.Nickname FROM {tableName} AS p");
            CountQueryResults(connectionString, $"SELECT p.PlayerId, p.Nickname, p.FirstRealPurchaseTime FROM {tableName} AS p");
            CountQueryResults(connectionString, $"SELECT p.Nickname, p.FirstRealPurchaseTime FROM {tableName} AS p");
            CountQueryResults(connectionString, $"SELECT p.FirstRealPurchaseTime FROM {tableName} AS p");
            CountQueryResults(connectionString, $"SELECT p.Nickname FROM {tableName} AS p");


**Results:**

09:59:18Z Counting results for SELECT p.PlayerId FROM Players AS p
10:00:16Z Total rows: 476577

10:00:16Z Counting results for SELECT p.PlayerId, p.FirstRealPurchaseTime FROM Players AS p
10:01:11Z Total rows: 476577

10:01:11Z Counting results for SELECT p.PlayerId, p.Nickname FROM Players AS p
10:02:05Z Total rows: 476577

10:02:05Z Counting results for SELECT p.PlayerId, p.Nickname, p.FirstRealPurchaseTime FROM Players AS p
**10:03:40Z Total rows: 452608**

10:03:40Z Counting results for SELECT p.Nickname, p.FirstRealPurchaseTime FROM Players AS p
**Null Pointer Exception**
   at Google.Cloud.Spanner.V1.ReliableStreamReader.MergeChunk(Value currentValue, Value nextValue)
   at Google.Cloud.Spanner.V1.ReliableStreamReader.<NextAsync>d__40.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.SpannerDataReader.<>c__DisplayClass57_0.<<ReadAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.ExecuteHelper.<WithErrorTranslationAndProfiling>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Google.Api.Gax.TaskExtensions.WaitWithUnwrappedExceptions(Task task)
   at Google.Cloud.Spanner.Data.SpannerDataReader.Read()
   at SpannerCleanTestingConsole.Program.CountQueryResults(String connectionString, String sql) in C:\Users\alyaros\Downloads\SpannerCleanTestingConsole\SpannerCleanTestingConsole\Program.cs:line 254
   at SpannerCleanTestingConsole.Program.Main(String[] args) in C:\Users\alyaros\Downloads\SpannerCleanTestingConsole\SpannerCleanTestingConsole\Program.cs:line 144

**Comment by alyaros** [comment]
Those also works correctly:
10:16:02Z Counting results for SELECT p.FirstRealPurchaseTime FROM Players AS p
10:17:03Z Total rows: 476577

10:17:03Z Counting results for SELECT p.Nickname FROM Players AS p
10:18:00Z Total rows: 476577

**Comment by alyaros** [comment]
Field Scheme Type:
Nickname | STRING(50) | Yes

**Comment by jskeet** [comment]
That's very interesting!

I suspect at this point the problem is very data-centric - and may well depend on other fields that aren't even being queried. I'm not sure of the best way to diagnose this.
If I were to create a special NuGet package file (.nupkg) for you to put on your local system in a directory that you add as a package source, would you be able to run the same test with that? I can add lots of extra logging, then you could email me the log file.

Let me know if that sounds feasible, and I'll start working on the extra logging.

**Comment by alyaros** [comment]
I ran the queries in different order - different results:
10:27:28Z Counting results for SELECT p.Nickname, p.FirstRealPurchaseTime FROM Players AS p
10:28:40Z Total rows: ****459716 (No NPE this time..)**

10:28:46Z Counting results for SELECT p.PlayerId, p.Nickname, p.FirstRealPurchaseTime FROM Players AS p
10:30:22Z Total rows: 452608

**Comment by alyaros** [comment]
I think I can send you privately those 3 columns dataset (PlayerId, FirstRealPurchaseTime, Nickname) - if this will make it easier\quicker for you to debug it.

**Comment by jskeet** [comment]
Let's try the logging first, but then that would be very helpful if that doesn't sort things. I wouldn't be surprised if even with that data, we found it only reproduced in certain situations - I'm going to be fascinated to see what the result of this is in the end...
I'm just testing my new packages locally first, then I'll put them on Google Drive.

**Comment by alyaros** [comment]
I ran it on a new table with only those 3 columns (p.PlayerId, p.Nickname, p.FirstRealPurchaseTime) and it does not reproduce, currently the only way to reproduce it is directly on this specific table.

**Comment by alyaros** [comment]
By the way the NPE is not stable - sometimes I get NPE and sometimes not.
10:27:28Z Counting results for SELECT p.Nickname, p.FirstRealPurchaseTime FROM Players AS p
10:28:40Z Total rows: **459716 (Sometimes NPE sometimes NOT )

10:28:46Z Counting results for SELECT p.PlayerId, p.Nickname, p.FirstRealPurchaseTime FROM Players AS p
10:30:22Z Total rows: 452608

**Comment by alyaros** [comment]
Different run different results:
10:50:04Z Counting results for SELECT p.PlayerId, p.Nickname, p.FirstRealPurchaseTime FROM Players AS p
10:52:53Z Total rows: **669583**

**Comment by jskeet** [comment]
Okay, I've attached the packages here as a zip file.

So you'll need to:

- Create a local directory (e.g. "LocalPackages")
- Unzip the zip file there
- Add that as a package source (e.g. in VS it's under "NuGet Package Manager / Package Sources" in settings)
- Update the dependencies in the project to 2.0.0-issue2779a

Then add the following line at the start of `Main`:

```csharp
Google.Cloud.Spanner.V1.Internal.Logging.Logger.DefaultLogger.LogLevel = Google.Cloud.Spanner.V1.Internal.Logging.LogLevel.Debug;
```

Run the program as before, redirecting the output to a text file. If you could then email me (jonskeet@google.com) with the log file, that would be wonderful. (You'll want to zip it first...)

[packages.zip](https://github.com/googleapis/google-cloud-dotnet/files/2723447/packages.zip)

**Comment by jskeet** [comment]
Okay, I'm pretty sure I now understand the issue.

The current default timeout is 1 minute. That's an RPC timeout; a query can last longer than that, but each call within the query times out after a minute. It then resumes from where it was - but the current implementation skips one result from the stream. That usually leads to missing rows; it could very occasionally lead to a NullReferenceException as per #2777. (I haven't observed that yet, but I can see how it *could* happen.)

Steps to fix:

- Test against 2.0.0-beta03 to find out why it wasn't failing there. My suspicion is that the default timeout is longer, but I need to confirm that. (I want to reproduce it failing with a 1 minute timeout, too.)
- Write unit tests for ReliableStreamReader. (Currently we just rely on integration tests which clearly don't cut it.)
- Optionally refactor the code, as it's fiddly at the moment.
- Remove the "extra MoveNext" call
- Potentially restrict the exceptions we retry on; I've seen one case where retrying after the gRPC stream was removed still caused a problem.
- Release

I'm going to see how much of that I can do today, although we probably wouldn't want to actually do the release on a Friday. I'm away next week, but may still be able to spend some time working on this - it's definitely my top priority.

**Comment by jskeet** [comment]
Closing as this is fixed in 2.0.0-beta06.
